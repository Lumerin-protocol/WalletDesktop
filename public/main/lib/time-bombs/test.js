const timeBombs = require('.')

const start = Date.now()

const bomb1 = timeBombs.create(250)

bomb1.onBoom(function () {
  console.log('Should show ~550ms:', Date.now() - start)
})

setTimeout(function () {
  bomb1.reset(200)
  setTimeout(function () {
    bomb1.reset(200)
  }, 150)
}, 200)

const bomb2 = timeBombs.create(100)

bomb2.onBoom(function () {
  console.log('Should show ~350ms:', Date.now() - start)
})

setTimeout(function () {
  bomb2.reset(300)
}, 50)
