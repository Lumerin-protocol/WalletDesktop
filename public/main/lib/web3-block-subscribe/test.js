const url = 'ws://parity.bloqrock.net:8546'
// const url = 'ws://localhost:8546'

const web3Subscription = require('.')

web3Subscription.subscribe({
  url,
  onData: function ({ number }) {
    console.log('Hash', number)
  },
  onError: function (err) {
    console.warn('Error', err.message)
  }
})
