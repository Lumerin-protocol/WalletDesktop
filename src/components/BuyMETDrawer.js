import { DisplayValue, Drawer, Btn, Sp } from './common'
import ConfirmationWizard from './ConfirmationWizard'
import * as validators from '../validator'
import * as selectors from '../selectors'
import AmountFields from './AmountFields'
import { debounce } from 'lodash'
import { connect } from 'react-redux'
import * as utils from '../utils'
import GasEditor from './GasEditor'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import config from '../config'
import React from 'react'
import Web3 from 'web3'

const ConfirmationContainer = styled.div`
  font-size: 1.3rem;
  font-weight: 600;
  letter-spacing: 0.5px;

  & > div {
    color: ${p => p.theme.colors.primary};
  }
`

const ExpectedMsg = styled.div`
  font-size: 1.3rem;
  color: ${p => (p.error ? p.theme.colors.danger : 'inherit')};
`

const BtnContainer = styled.div`
  background-image: linear-gradient(to bottom, #272727, #323232);
  padding: 3.2rem 2.4rem;
`

class BuyMETDrawer extends React.Component {
  static propTypes = {
    tokenRemaining: PropTypes.string.isRequired,
    onRequestClose: PropTypes.func.isRequired,
    currentPrice: PropTypes.string.isRequired,
    availableETH: PropTypes.string.isRequired,
    ETHprice: PropTypes.number.isRequired,
    isOpen: PropTypes.bool.isRequired,
    from: PropTypes.string.isRequired
  }

  static initialState = {
    ...AmountFields.initialState,
    ...GasEditor.initialState('MET'),
    errors: {}
  }

  state = BuyMETDrawer.initialState

  componentWillReceiveProps(newProps) {
    if (newProps.isOpen && newProps.isOpen !== this.props.isOpen) {
      this.setState(BuyMETDrawer.initialState)
    }
  }

  onInputChange = e => {
    const { id, value } = e.target
    const { ETHprice } = this.props

    this.setState(state => ({
      ...state,
      ...AmountFields.onInputChange(state, ETHprice, id, value),
      errors: { ...state.errors, [id]: null },
      [id]: value
    }))

    // Estimate gas limit again if parameters changed
    if (['ethAmount'].includes(id)) this.getGasEstimate()
  }

  getGasEstimate = debounce(() => {
    const { ethAmount } = this.state

    if (!utils.isWeiable(ethAmount)) return

    utils
      .sendToMainProcess('metronome-auction-gas-limit', {
        value: Web3.utils.toWei(ethAmount.replace(',', '.')),
        from: this.props.from
      })
      .then(({ gasLimit }) => this.setState({ gasLimit: gasLimit.toString() }))
      .catch(err => console.warn('Gas estimation failed', err))
  }, 500)

  onWizardSubmit = password => {
    return utils.sendToMainProcess('metronome-buy', {
      gasPrice: Web3.utils.toWei(this.state.gasPrice.replace(',', '.'), 'gwei'),
      gasLimit: this.state.gasLimit,
      password,
      value: Web3.utils.toWei(this.state.ethAmount.replace(',', '.')),
      from: this.props.from
    })
  }

  validate = () => {
    const { ethAmount, gasPrice, gasLimit } = this.state
    const max = Web3.utils.fromWei(this.props.availableETH)
    const errors = {
      ...validators.validateEthAmount(ethAmount, max),
      ...validators.validateGasPrice(gasPrice, config.MAX_GAS_PRICE),
      ...validators.validateGasLimit(gasLimit)
    }
    const hasErrors = Object.keys(errors).length > 0
    if (hasErrors) this.setState({ errors })
    return !hasErrors
  }

  renderConfirmation = () => {
    const { currentPrice, tokenRemaining } = this.props
    const { ethAmount, usdAmount } = this.state

    const expected = utils.toMET(ethAmount, currentPrice, null, tokenRemaining)

    return (
      <ConfirmationContainer data-testid="confirmation">
        {expected.excedes ? (
          <React.Fragment>
            You will use{' '}
            <DisplayValue value={expected.usedETHAmount} post=" ETH" inline />{' '}
            to buy{' '}
            <DisplayValue
              inline
              value={this.props.tokenRemaining}
              post=" MET"
            />{' '}
            at current price and get a return of approximately{' '}
            <DisplayValue inline value={expected.excessETHAmount} post=" ETH" />.
            <Sp my={2}>
              <ExpectedMsg error>
                This operation will deplete the current auction.
              </ExpectedMsg>
            </Sp>
          </React.Fragment>
        ) : (
          <React.Fragment>
            You will use{' '}
            <DisplayValue
              inline
              value={Web3.utils.toWei(ethAmount.replace(',', '.'))}
              post=" ETH"
            />{' '}
            (${usdAmount}) to buy approximately{' '}
            <DisplayValue
              inline
              value={expected.expectedMETamount}
              post=" MET"
            />{' '}
            at current price.
          </React.Fragment>
        )}
      </ConfirmationContainer>
    )
  }

  renderForm = goToReview => {
    const {
      useCustomGas,
      ethAmount,
      usdAmount,
      gasPrice,
      gasLimit,
      errors
    } = this.state

    const { expectedMETamount, excedes, excessETHAmount } = utils.toMET(
      ethAmount,
      this.props.currentPrice,
      null,
      this.props.tokenRemaining
    )

    return (
      <form onSubmit={goToReview} noValidate data-testid="buy-form">
        <Sp py={4} px={3}>
          <AmountFields
            availableETH={this.props.availableETH}
            ethAmount={ethAmount}
            usdAmount={usdAmount}
            autoFocus
            onChange={this.onInputChange}
            errors={errors}
          />

          <Sp mt={3}>
            <GasEditor
              useCustomGas={useCustomGas}
              onChange={this.onInputChange}
              gasPrice={gasPrice}
              gasLimit={gasLimit}
              errors={errors}
            />
          </Sp>

          {expectedMETamount && (
            <Sp mt={2}>
              {excedes ? (
                <ExpectedMsg error>
                  You would get all remaining{' '}
                  <DisplayValue
                    inline
                    value={this.props.tokenRemaining}
                    post=" MET"
                  />{' '}
                  and receive a return of approximately{' '}
                  <DisplayValue inline value={excessETHAmount} post=" ETH" />.
                </ExpectedMsg>
              ) : (
                <ExpectedMsg>
                  You would get approximately{' '}
                  <DisplayValue inline value={expectedMETamount} post=" MET" />.
                </ExpectedMsg>
              )}
            </Sp>
          )}
        </Sp>

        <BtnContainer>
          <Btn submit block>
            Review Buy
          </Btn>
        </BtnContainer>
      </form>
    )
  }

  render() {
    const { onRequestClose, isOpen } = this.props

    return (
      <Drawer
        onRequestClose={onRequestClose}
        data-testid="buy-drawer"
        isOpen={isOpen}
        title="Buy Metronome"
      >
        <ConfirmationWizard
          renderConfirmation={this.renderConfirmation}
          onWizardSubmit={this.onWizardSubmit}
          pendingTitle="Buying MET..."
          renderForm={this.renderForm}
          editLabel="Edit this buy"
          validate={this.validate}
        />
      </Drawer>
    )
  }
}

const mapStateToProps = state => ({
  availableETH: selectors.getEthBalanceWei(state),
  ETHprice: selectors.getEthRate(state),
  from: selectors.getActiveWalletAddresses(state)[0]
})

export default connect(mapStateToProps)(BuyMETDrawer)
