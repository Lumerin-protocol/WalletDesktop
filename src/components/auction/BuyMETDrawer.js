import withBuyMETFormState from 'metronome-wallet-ui-logic/src/hocs/withBuyMETFormState'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import React from 'react'

import {
  ConfirmationWizard,
  DisplayValue,
  AmountFields,
  GasEditor,
  Drawer,
  Btn,
  Sp
} from '../common'

const ConfirmationContainer = styled.div`
  font-size: 1.3rem;
  font-weight: 600;
  letter-spacing: 0.5px;

  & > div {
    color: ${p => p.theme.colors.primary};
  }
`

const ExpectedMsg = styled.div`
  font-size: 1.3rem;
  color: ${p => (p.error ? p.theme.colors.danger : 'inherit')};
`

const BtnContainer = styled.div`
  background-image: linear-gradient(to bottom, #272727, #323232);
  padding: 3.2rem 2.4rem;
`

class BuyMETDrawer extends React.Component {
  static propTypes = {
    expectedMETamount: PropTypes.string,
    gasEstimateError: PropTypes.bool,
    excessCoinAmount: PropTypes.string,
    onRequestClose: PropTypes.func.isRequired,
    coinPlaceholder: PropTypes.string,
    usdPlaceholder: PropTypes.string,
    tokenRemaining: PropTypes.string,
    onInputChange: PropTypes.func.isRequired,
    usedCoinAmount: PropTypes.string,
    useCustomGas: PropTypes.bool.isRequired,
    onMaxClick: PropTypes.func.isRequired,
    coinSymbol: PropTypes.string.isRequired,
    coinAmount: PropTypes.string,
    resetForm: PropTypes.func.isRequired,
    usdAmount: PropTypes.string,
    onSubmit: PropTypes.func.isRequired,
    validate: PropTypes.func.isRequired,
    gasPrice: PropTypes.string,
    gasLimit: PropTypes.string,
    excedes: PropTypes.bool,
    errors: PropTypes.object.isRequired,
    isOpen: PropTypes.bool.isRequired
  }

  componentDidUpdate(prevProps) {
    if (prevProps.isOpen && prevProps.isOpen !== this.props.isOpen) {
      this.props.resetForm()
    }
  }

  renderConfirmation = () => (
    <ConfirmationContainer data-testid="confirmation">
      {this.props.excedes ? (
        <React.Fragment>
          You will use{' '}
          <DisplayValue inline value={this.props.usedCoinAmount} isCoin /> to
          buy{' '}
          <DisplayValue inline value={this.props.tokenRemaining} post=" MET" />{' '}
          at current price and get a return of approximately{' '}
          <DisplayValue inline isCoin value={this.props.excessCoinAmount} />.
          <Sp my={2}>
            <ExpectedMsg error>
              This operation will deplete the current auction.
            </ExpectedMsg>
          </Sp>
        </React.Fragment>
      ) : (
        <React.Fragment>
          You will use{' '}
          <DisplayValue inline value={this.props.coinAmount} toWei isCoin /> (
          {this.props.usdAmount ? `$${this.props.usdAmount}` : '< $0.01'}) to
          buy approximately{' '}
          <DisplayValue
            inline
            value={this.props.expectedMETamount}
            post=" MET"
          />{' '}
          at current price.
        </React.Fragment>
      )}
    </ConfirmationContainer>
  )

  renderForm = goToReview => (
    <form onSubmit={goToReview} noValidate data-testid="buy-form">
      <Sp py={4} px={3}>
        <AmountFields
          coinPlaceholder={this.props.coinPlaceholder}
          usdPlaceholder={this.props.usdPlaceholder}
          onMaxClick={this.props.onMaxClick}
          coinAmount={this.props.coinAmount}
          coinSymbol={this.props.coinSymbol}
          usdAmount={this.props.usdAmount}
          autoFocus
          onChange={this.props.onInputChange}
          errors={this.props.errors}
        />

        <Sp mt={3}>
          <GasEditor
            gasEstimateError={this.props.gasEstimateError}
            onInputChange={this.props.onInputChange}
            useCustomGas={this.props.useCustomGas}
            gasLimit={this.props.gasLimit}
            gasPrice={this.props.gasPrice}
            errors={this.props.errors}
          />
        </Sp>

        {this.props.expectedMETamount && (
          <Sp mt={2}>
            {this.props.excedes ? (
              <ExpectedMsg error>
                You would get all remaining{' '}
                <DisplayValue
                  inline
                  value={this.props.tokenRemaining}
                  post=" MET"
                />{' '}
                and receive a return of approximately{' '}
                <DisplayValue
                  inline
                  isCoin
                  value={this.props.excessCoinAmount}
                />
                .
              </ExpectedMsg>
            ) : (
              <ExpectedMsg>
                You would get approximately{' '}
                <DisplayValue
                  inline
                  value={this.props.expectedMETamount}
                  post=" MET"
                />
                .
              </ExpectedMsg>
            )}
          </Sp>
        )}
      </Sp>

      <BtnContainer>
        <Btn submit block>
          Review Buy
        </Btn>
      </BtnContainer>
    </form>
  )

  render() {
    return (
      <Drawer
        onRequestClose={this.props.onRequestClose}
        data-testid="buy-drawer"
        isOpen={this.props.isOpen}
        title="Buy Metronome"
      >
        <ConfirmationWizard
          renderConfirmation={this.renderConfirmation}
          onWizardSubmit={this.props.onSubmit}
          pendingTitle="Buying MET..."
          renderForm={this.renderForm}
          editLabel="Edit this buy"
          validate={this.props.validate}
        />
      </Drawer>
    )
  }
}

export default withBuyMETFormState(BuyMETDrawer)
