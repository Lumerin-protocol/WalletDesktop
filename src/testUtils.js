/* eslint-disable require-path-exists/exists */
/* eslint-disable import/no-unresolved */
import { Provider as ClientProvider } from 'lumerin-wallet-ui-logic/src/hocs/clientContext';
import { Provider, createStore } from 'lumerin-wallet-ui-logic/src/store';
import { render, Simulate } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router';
import createClient from './client';
import { merge } from 'lodash';
import theme from 'lumerin-wallet-ui-logic/src/theme';
import React from 'react';

import config from '../config';

/*
 * The same render method of 'react-testing-library' but wrapped with
 * Redux store Provider and 'styled-components' ThemeProvider.
 *
 * element      : the React element you want to wrap
 * initialState : optional state to initialize the Redux store
 *
 * Returns the same object as render() but with an additional 'store'
 * property useful for dispatching actions inside your tests.
 */
export function reduxRender(element, initialState) {
  const client = createClient(config, createStore, initialState);

  const renderResult = render(
    <ClientProvider value={client}>
      <Provider store={client.store}>
        <ThemeProvider theme={theme}>
          <React.Fragment>{element}</React.Fragment>
        </ThemeProvider>
      </Provider>
    </ClientProvider>
  );
  return { ...renderResult, store: client.store };
}

/*
 * The same as reduxRender() above but also wrapped with 'react-router'
 * in order to test navigation flows.
 *
 * Returns the same object as reduxRender()
 */
export function routerRender(element, initialState, initialRoute = '/') {
  return reduxRender(
    <MemoryRouter initialEntries={[initialRoute]}>{element}</MemoryRouter>,
    initialState
  );
}

/*
 * Declaratively test complex form validations
 *
 * getter   : function that receives a data-testid and returns an HTML element
 * formId   : data-testid of the form element
 * formData : object of type { [data-testid]: any } to populate the form
 * errors   : object of type { [data-testid]: String } with expected errors
 */
export function testValidation(getter, formId, { formData, errors }) {
  // Populate the specified form fields
  Object.keys(formData).forEach((fieldId, i) => {
    const field = getter(fieldId);
    field.value = formData[fieldId];
    Simulate.change(field);
  });

  // Submit the form
  Simulate.submit(getter(formId));

  // Check expected errors
  Object.keys(errors).forEach((fieldId, i) => {
    expect(getter(`${fieldId}-error`).textContent).toEqual(
      expect.stringContaining(errors[fieldId])
    );
  });
}

/*
 * Add dataset property (data-foo attribute values) to nodes
 *
 * This hack is required because jsdom has no support for dataset yet
 * which prevents us from testing components that rely on it.
 * See https://github.com/jsdom/jsdom/issues/961
 *
 * element   : DOM node to add the dataset to
 * dataAttrs : attribute names to add to the node
 *
 * Usage:
 * <button data-testid="btn" data-modal="buy" data-foo="bar">Click</button>
 *
 * const btn = withDataset(getByTestId('btn'), 'modal', 'foo')
 * // btn.dataset === { modal: 'buy', foo: 'bar'}
 *
 */
export function withDataset(element, ...dataAttrs) {
  element.dataset = dataAttrs.reduce((acc, attr) => {
    acc[attr] = element.getAttribute(`data-${attr}`);
    return acc;
  }, {});
  return element;
}

/**
 * Return UTC timestamps relative to the current time
 * Useful for setting up different auction scenarios.
 */
export const twoWeeksAgo = () => Date.now() / 1000 - 60 * 60 * 24 * 7 * 2;
export const oneHourAgo = () => Date.now() / 1000 - 60 * 60;
export const inOneHour = () => Date.now() / 1000 + 60 * 60;
export const inOneWeek = () => Date.now() / 1000 + 60 * 60 * 24 * 7;

/*
 * Returns a common initial state for Redux store that is useful for most tests
 * Accepts an optional object with overrides that will be deeply merged with
 * the base state object.
 */
export function getInitialState(overrides = {}) {
  const baseState = {
    config,
    connectivity: { isOnline: true },
    blockchain: { height: 1, gasPrice: config.DEFAULT_GAS_PRICE },
    converter: {
      status: {
        availableEth: '100',
        availableLmr: '100',
        currentPrice: '10'
      }
    },
    auction: {
      status: {
        nextAuctionStartTime: inOneHour(),
        tokenRemaining: '1',
        currentAuction: 10,
        currentPrice: '33000000000',
        genesisTime: twoWeeksAgo()
      }
    },
    session: { isLoggedIn: true },
    rates: { ETH: { token: 'ETH', price: 1 } },
    wallets: {
      syncStatus: 'up-to-date',
      active: 'foo',
      allIds: ['foo'],
      byId: {
        foo: {
          addresses: {
            '0x15dd2028C976beaA6668E286b496A518F457b5Cf': {
              token: {
                [config.LMR_TOKEN_ADDR]: { balance: '5000000000000000000000' }
              },
              balance: '5000000000000000000000',
              transactions: []
            }
          }
        }
      }
    }
  };
  return merge({}, baseState, overrides);
}

/*
 * A sample transaction object. Useful for populating state in tests.
 */
export function getDummyTransaction() {
  return {
    meta: {
      lumerin: {},
      tokens: {
        [config.LMR_TOKEN_ADDR]: {
          event: 'Transfer',
          processing: false,
          value:
            '0x000000000000000000000000000000000000000000000001e5b8fa8fe2ac0000'
        }
      }
    },
    receipt: {
      blockHash: '0x1234567890abcdef',
      blockNumber: 407574,
      cumulativeGasUsed: 38484,
      gasUsed: 38484,
      transactionHash: '0x1234567890abcdef',
      transactionIndex: 0
    },
    transaction: {
      transactionIndex: 0,
      blockNumber: 407574,
      blockHash: '0x1234567890abcdef',
      hash: '0x1234567890abcdef',
      value: '0',
      gasPrice: '5000000000',
      gas: 76968,
      from: '0x15dd2028C976beaA6668E286b496A518F457b5Cf',
      to: '0xde806D6efD432CDeE42573760682D99eDEdC1d89'
    }
  };
}
