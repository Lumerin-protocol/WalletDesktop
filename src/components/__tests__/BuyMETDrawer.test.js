import * as amountFields from './AmountFields.test.js'
import * as gasEditor from './GasEditor.test.js'
import * as testUtils from '../../testUtils'
import { Simulate } from 'react-testing-library'
import BuyMETDrawer from '../BuyMETDrawer'
import React from 'react'

const closeHandler = jest.fn()

const element = (
  <BuyMETDrawer
    tokenRemaining="100"
    onRequestClose={closeHandler}
    currentPrice="10"
    isOpen
  />
)

const ETHprice = 250

describe('<BuyMETDrawer/>', () => {
  it('should match its snapshot', () => {
    const { container } = testUtils.reduxRender(element, getInitialState())
    expect(container).toMatchSnapshot()
  })

  describe('When editing the amount fields', () => {
    amountFields.runEditTests(element, getInitialState(), ETHprice)
  })

  describe('When submitting the form', () => {
    amountFields.runValidateTests(element, getInitialState(), 'buy-form')

    gasEditor.runValidateTests(element, getInitialState(), 'buy-form')

    it('displays the confirmation view if there are no errors', () => {
      const { queryByTestId, getByTestId } = testUtils.reduxRender(
        element,
        getInitialState()
      )
      expect(queryByTestId('confirmation')).toBeNull()
      const amountField = getByTestId('ethAmount-field')
      amountField.value = '1'
      Simulate.change(amountField)
      Simulate.submit(getByTestId('buy-form'))
      expect(queryByTestId('confirmation')).not.toBeNull()
    })
  })
})

function getInitialState() {
  return testUtils.getInitialState({
    rates: { ETH: { token: 'ETH', price: ETHprice } }
  })
}
