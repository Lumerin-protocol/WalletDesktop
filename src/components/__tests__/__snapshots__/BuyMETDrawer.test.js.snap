// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AmountFields/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-jKJlTe fkNCah"
  >
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <button
        type="button"
        class="sc-htpNat fpAZyb"
        data-testid="max-btn"
        tabindex="-1"
      >
        MAX
      </button>
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="ethAmount"
        >
          Amount (ETH)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="ethAmount"
          data-testid="ethAmount-field"
          placeholder="0.00"
          value=""
        />
      </div>
    </div>
    <div
      class="sc-hMqMXs dnTKXF"
    >
      <svg
        class="sc-bxivhb hXQRLM"
        viewBox="0 0 24 24"
      >
        <path
          d="M24 7.136v.96H1.642l4.464 4.464-.682.682-5.28-5.28a.475.475 0 0 1 0-.682L5.424 2l.682.682-4.464 4.464H24v-.01zM18.576 11.6l-.682.682 4.464 4.464H0v.96h22.358l-4.464 4.464.682.681 5.28-5.28a.475.475 0 0 0 0-.681l-5.28-5.29z"
        />
      </svg>
    </div>
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="usdAmount"
        >
          Amount (USD)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="usdAmount"
          data-testid="usdAmount-field"
          placeholder="0.00"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<BuyMETDrawer/> should match its snapshot 1`] = `
<div>
  <header
    class="sc-jTzLTM bbfTuz"
  >
    <h1
      class="sc-fjdhpX gUwcgC"
      data-testid="drawer-title"
    >
      Buy Metronome
    </h1>
    <button
      class="sc-jzJRlG fzwWXa"
      data-testid="drawer-close-btn"
    >
      <svg
        class="sc-bxivhb hXQRLM"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 24C5.386 24 0 18.614 0 12S5.386 0 12 0s12 5.386 12 12-5.386 12-12 12zM12 .96C5.914.96.96 5.914.96 12c0 6.086 4.954 11.04 11.04 11.04 6.086 0 11.04-4.954 11.04-11.04C23.04 5.914 18.086.96 12 .96zm3.984 15.696L12 12.672l-3.984 3.984-.682-.682 3.984-3.984-3.984-3.984.682-.681L12 11.309l3.984-3.984.682.681-3.984 3.984 3.984 3.984-.682.682z"
        />
      </svg>
    </button>
  </header>
  <div
    data-testid="buy-drawer"
  >
    <form
      novalidate=""
      data-testid="buy-form"
    >
      <div
        class="sc-hMqMXs dKMJsu"
      >
        <div
          class="sc-jKJlTe fkNCah"
        >
          <div
            class="sc-ckVGcZ kvHIve"
          >
            <button
              type="button"
              class="sc-htpNat fpAZyb"
              data-testid="max-btn"
              tabindex="-1"
            >
              MAX
            </button>
            <div>
              <label
                class="sc-dnqmqq ilsafR"
                for="ethAmount"
              >
                Amount (ETH)
              </label>
              <input
                class="sc-iwsKbI kZiHEX"
                type="text"
                id="ethAmount"
                data-testid="ethAmount-field"
                placeholder="0.00"
                value=""
              />
            </div>
          </div>
          <div
            class="sc-hMqMXs dnTKXF"
          >
            <svg
              class="sc-bxivhb hXQRLM"
              viewBox="0 0 24 24"
            >
              <path
                d="M24 7.136v.96H1.642l4.464 4.464-.682.682-5.28-5.28a.475.475 0 0 1 0-.682L5.424 2l.682.682-4.464 4.464H24v-.01zM18.576 11.6l-.682.682 4.464 4.464H0v.96h22.358l-4.464 4.464.682.681 5.28-5.28a.475.475 0 0 0 0-.681l-5.28-5.29z"
              />
            </svg>
          </div>
          <div
            class="sc-ckVGcZ kvHIve"
          >
            <div>
              <label
                class="sc-dnqmqq ilsafR"
                for="usdAmount"
              >
                Amount (USD)
              </label>
              <input
                class="sc-iwsKbI kZiHEX"
                type="text"
                id="usdAmount"
                data-testid="usdAmount-field"
                placeholder="0.00"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="sc-hMqMXs bQjXee"
        >
          <div
            class="sc-jKJlTe kWLJKD"
          >
            <div
              class="sc-jKJlTe jGctBw"
            >
              <span
                class="sc-kEYyzF ekxbwv"
              >
                Gas Limit: 
                250000
                (Units)
              </span>
              <span
                class="sc-kEYyzF ekxbwv"
              >
                Gas Price: 
                1
                (Gwei)
              </span>
            </div>
            <button
              type="button"
              class="sc-htpNat bqwpzm"
              data-testid="edit-gas-btn"
              tabindex="-1"
            >
              EDIT GAS
            </button>
          </div>
        </div>
      </div>
      <div
        class="sc-gPEVay fwZOlX"
      >
        <button
          type="submit"
          class="sc-bwzfXH gAMukr"
        >
          Review Buy
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`<GasEditor/> should match its snapshot 1`] = `
<div>
  <div
    class="sc-jKJlTe fkNCah"
  >
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="gasLimit"
        >
          Gas Limit (Units)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="gasLimit"
          data-testid="gas-limit-field"
          value="21000"
        />
      </div>
    </div>
    <div
      class="sc-hMqMXs keWZLk"
    />
    <div
      class="sc-ckVGcZ kvHIve"
    >
      <div>
        <label
          class="sc-dnqmqq ilsafR"
          for="gasPrice"
        >
          Gas Price (Gwei)
        </label>
        <input
          class="sc-iwsKbI kZiHEX"
          type="text"
          id="gasPrice"
          data-testid="gas-price-field"
          value="1"
        />
      </div>
    </div>
  </div>
</div>
`;
